### ISSUE TITLE
Subclasses of sealed classes are not registered 

### ISSUE THREAD

**Issue by TimerErTim** [Kotlin/kotlinx.serialization#1854]
**Describe the bug**

A simple polymorphic serialization setup using sealed classes (like in the official kotlinx serialization guide) causes an exception as if the subclass was not registered for polymorphic serialization in the scope of the superclass. 

**To Reproduce**
``` kotlin
@Serializable
sealed class Fetcher {
    abstract suspend fun fetch(url: URL): URL
}

@Serializable
@SerialName("Plain")
object PlainFetcher : Fetcher() {
    override suspend fun fetch(url: URL) = url
}

fun main() {
    val fetcher: Fetcher = PlainFetcher
    println(Json.encodeToString(fetcher))
}
```

Running the program results in following error:

> kotlinx.serialization.SerializationException: Class 'PlainFetcher' is not registered for polymorphic serialization in the scope of 'Fetcher'.
	Mark the base class as 'sealed' or register the serializer explicitly.

**NOTE**:  The first time I compiled the example the serialization worked pretty much flawlessly. Only after I had made some changes (to the rest of the code resulting in recompilation of the module) and recompiled, the error occurred. Restoring the code to the state of the first compilation still resulted in an error when recompiled.

**Expected behavior**

No error and successful serialization of the fetcher.

**Environment**
 - Kotlin version: 1.6.10
 - Library version: 1.3.2
 - Kotlin platforms: JVM
 - Gradle version: 7.2
 - IDE version (if bug is related to the IDE): IntelliJ IDEA 2021.3.2 Ultimate
 - Other relevant context: OpenSUSE Tumbleweed 64-Bit,  OpenJDK 11.0.10+9

**Comment by TimerErTim** [comment]
The problem is gone when the inheriting class is put in the same file as the sealed class. This is basically a duplicate of #1541. Based on the comment history however, #1541 seems to have been resolved. This issue shows: that is not the case.
